### DISCUSSION TITLE
Experimental Agents broken with custom endpoints

### DISCUSSION THREAD

**Discussion by returnofblank** [danny-avila/LibreChat#4303] [Troubleshooting]
### What happened?

Note: This is only tested with OpenRouter, this may not apply to other custom endpoints. Unfortunately I do not have the resources to try with something else.

When choosing a custom endpoint in the agent creation menu, and then choosing a model (any model from what I found), exiting out will lead to this error.

```
Unexpected Application Error!
n.toLowerCase is not a function
i_@https://mydomain.example/assets/index.DOfEXLPm.js:3128:22877
A7@https://mydomain.example/assets/vendor.BujiSg_e.js:38:19557
Qee@https://mydomain.example/assets/vendor.BujiSg_e.js:40:44171
zee@https://mydomain.example/assets/vendor.BujiSg_e.js:40:39879
pSe@https://mydomain.example/assets/vendor.BujiSg_e.js:40:39802
v2@https://mydomain.example/assets/vendor.BujiSg_e.js:40:39648
j$@https://mydomain.example/assets/vendor.BujiSg_e.js:40:36004
iB@https://mydomain.example/assets/vendor.BujiSg_e.js:40:36809
Kf@https://mydomain.example/assets/vendor.BujiSg_e.js:38:3288
Vo/<@https://mydomain.example/assets/vendor.BujiSg_e.js:40:34324
```

I have no concrete theory as to what is causing this. But I assume something on the front end is passing a broken input to the `chatCompletion` function at `api/server/controllers/agents/client.js` since the error appears in the logs.

Again, this theory is mostly baseless. I unfortunately do not have the time to go more in depth, as I wish to rewatch The Orville at this moment, not make another issue.

### Steps to Reproduce

1. Create a custom endpoint (for me, OpenRouter)
2. Under the Model option, choose the custom endpoint provider
3. Set the actual model to anything, not sure it matters
4. Click the back arrow to return to the previous screen
5. Error will appear

### Relevant log output

```shell
2024-10-02T02:03:46.493Z debug: [api/server/controllers/agents/client.js] chatCompletion
{
  baseURL: undefined,
  // 1 payload(s)
  payload: [{"role":"user","content":"\nUnexpected Application Error!\nn.toLowerCase is not a function\ni_@ht... [truncated]],
}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by returnofblank** [comment]
I also didn't see this when I first made the issue, but someone on the discord is also experiencing the issue.

So it's not an isolated incident with my f*cked up config at least.

![image](https://github.com/user-attachments/assets/cfc48093-1fac-47d7-a736-244b3eac6f44)

**Comment by bsu3338** [comment]
I will confirm that I had the issue. I use custom and litellm as a backend.

**Comment by danny-avila** [comment]
This is expected since functionality with custom endpoints is not currently supported.

I have not advertised this feature to be used. If you find it great, you're welcome to try it out and report issues, I may even start a dedicated discussion for this once I make a proper announcement for users to test this.

**Comment by hanna-daoud** [comment]
I also had this issue, here is a more relevant screenshot where I ran it on the dev endpoint. Note that it called my endpoint successfully, it just throws this in the end. 

![image](https://github.com/user-attachments/assets/48ef4d2c-3226-46a8-a9fb-a2105e1fb32b)

**Comment by hanna-daoud** [comment]
I made this PR #4324 as a fix.

**Comment by NTNguyen13** [comment]
in my case, I don't see custom endpoint available under Provider:
![image](https://github.com/user-attachments/assets/670be1cb-9793-40d6-8400-07e33b4981b3)
I added these to librechat.yml
```
  agents:
    # (optional) Maximum recursion depth for agents, defaults to 25
    recursionLimit: 50
    # (optional) Disable the builder interface for agents
    disableBuilder: false
    # (optional) Agent Capabilities available to all users. Omit the ones you wish to exclude. Defaults to list below.
    capabilities: ["execute_code", "file_search", "actions", "tools"]
```

and this to .env
`ENDPOINTS=openAI,assistants,azureOpenAI,bingAI,google,gptPlugins,anthropic,custom,agents
`

**Reply by NTNguyen13** [reply]
for some reasons, it's working now
